core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnEncoding.scala

```diff
@@ -304,15 +304,16 @@ trait ColumnEncoder extends ColumnEncoding {
     else if (numNullWords != 0) assert(assertion = false,
       s"Unexpected nulls=$numNullWords for withHeader=false")
 
-    if (columnData eq null) {
-      var initByteSize: Long = 0L
-      if (reuseUsedSize > 0) {
+    var baseSize: Long = numNullBytes
+    if (withHeader) {
+      baseSize += 8L /* typeId + nullsSize */
+    }
+    if ((columnData eq null) || (columnData.limit() < (baseSize + defSize))) {
+      var initByteSize = 0L
+      if (reuseUsedSize > baseSize + defSize) {
         initByteSize = reuseUsedSize
       } else {
-        initByteSize = defSize.toLong * initSize + numNullBytes
-        if (withHeader) {
-          initByteSize += 8L /* typeId + nullsSize */
-        }
+        initByteSize = defSize.toLong * initSize + baseSize
```

Comment:
The baseSize already has the "numNullBytes" + header of 8 bytes, so removed from here and added baseSize.